universe/ValueRef.h

```diff
@@ -84,4 +90,96 @@ enum StatisticType : int {
+    //! free function GetValueRef(...) instead, mainly to save some typing.
+    auto GetValueRefBase(const std::string& name) const -> ValueRef::ValueRefBase* const;
+
+    /** returns a map with all named value refs */
+    auto GetItems() const -> any_container_type;
+
+    // Singleton
+    NamedValueRefManager&  operator=(NamedValueRefManager const&) = delete; // no copy via assignment
+
+    NamedValueRefManager(NamedValueRefManager const&) = delete;            // no copies via construction
+
+    ~NamedValueRefManager();
+
+    //! Returns the instance of this singleton class; you should use the free
+    //! function GetNamedValueRefManager() instead
+    static NamedValueRefManager& GetNamedValueRefManager();
+
+    //! Returns a number, calculated from the contained data, which should be
+    //! different for different contained data, and must be the same for
+    //! the same contained data, and must be the same on different platforms
+    //! and executions of the program and the function. Useful to verify that
+    //! the parsed content is consistent without sending it all between
+    //! clients and server.
+    auto GetCheckSum() const -> unsigned int;
+
+    //! Register the @p value_ref under the evaluated @p name.
+    template <typename T>
+    void RegisterValueRef(std::string&& name, std::unique_ptr<ValueRef::ValueRef<T>>&& vref);
+
+private:
+    NamedValueRefManager();
+
+    /** helper function for GetValueRef */
+    template <typename V>
+    V* const GetValueRefImpl(std::map<key_type, std::unique_ptr<V>>& registry, const std::string& label, const std::string& name);
+
+    /** helper function for RegisterValueRef */
+    template <typename R, typename VR>
+    void RegisterValueRefImpl(R& registry, std::mutex& mutex, const std::string& label, std::string&& valueref_name, std::unique_ptr<VR>&& vref);
+
+    //! Map of ValueRef%s identified by a name and mutexes for those to allow asynchronous registration
+    double_container_type m_value_refs_double; // int value refs
+    std::mutex            m_value_refs_double_mutex;
+    int_container_type    m_value_refs_int; // int value refs
+    std::mutex            m_value_refs_int_mutex;
+    container_type        m_value_refs; // everything else
+    std::mutex            m_value_refs_mutex;
+
+    // The s_instance creation is lazily triggered via a function local.
+    // There is exactly one for all translation units.
+    static NamedValueRefManager* s_instance;
+};
+
+//! Returns the singleton manager for named value refs
+FO_COMMON_API auto GetNamedValueRefManager() -> NamedValueRefManager&;
+
+//! Returns the ValueRef object registered with the given
+//! @p name.  If no such ValueRef exists, nullptr is returned instead.
+FO_COMMON_API auto GetValueRefBase(const std::string& name) -> ValueRef::ValueRefBase* const;
+
+template <typename T>
+FO_COMMON_API auto GetValueRef(const std::string& name) -> ValueRef::ValueRef<T>* const;
```

Comment:
Declaring template functions in a header but not defining or declaring specializations doesn't make much sense... Can remove these declarations since the only use appears to be in NamedValueRefManager.cpp

Reply:
`::GetValueRef<T>(m_value_ref_name)` is also used in `NamedRef<T>::GetValueRef()` in ValueRefs.h

Reply:
...and `::GetValueRefBase` is used in the UI/LinkText (maybe also VarText)

Reply:
I guess if you explicitly instantiate `GetValueRef<T>` for all the needed types in one .cpp, it will work without. Somewhat awkward, though... eg. I think presently there won't be a `std::string` version available, since it's not explicitly instantiated...?

`GetValueRefBase` isn't templated so not what I'm talking about

Reply:
> Declaring template functions in a header but not defining or declaring specializations doesn't make much sense... 

I thought using the functions in the parser would trigger instantiation for the actually necessary types but it seems it does not work this way. maybe the instantiation ends up in the wrong translation unit?

Reply:
If you don't have the template definitions in the header, then they can't be instantiated anywhere other than the TU where they are defined.

Reply:
> > I think putting the value ref registration/lookup code (NamedValueRefManager) makes about as much sense as putting the basic ValueRef definition in ValueRefs.
> 
> Not sure what you mean by that...

I meant that if splitting ValueRef.h from ValueRefs.h makes sense because else the declaration gets swamped by subclass declartions, it makes sense to split the registration/lookup code so it does not get swamped by subclass declarations.

> 
> > So maybe rather a NamedValueRefManager.h ?
> 
> Could be reasonable.
> 
> How many places actually need to access the stuff that would be in it? Most places using value refs probably don't directly... just when evaluating a value ref that includes a named ref... so maybe it would just be included in ValueRefs.cpp and some parsers?

Registration only happens in (some of the) parsers. Lookup only happens in LinkText/VarText and NamedValueRef (which is in ValueRefs.cpp). So changes to NamedValueRefManager.h would hopefully not trigger a big rebuild(?)

> 
> > If we go for non-template overloads would it not also make sense to go for GetIntValueRef, GetDoubleValueRef?
> 
> Possibly...
> 

yikes. did not think you would think that is a good idea.

> > Or do something with heterogenous containers using boost hana?
> 
> Not sure what that is, but it sounds unnecessarily complicated.

That is a boost library based on c++14 for metaprogramming which also supports heterogenous containers. Probably  I could use a single map containing member for ValueRef<int>, ValueRef<double>, ...